Implement Access.at/1

amuino opened the pull request:
It was described, but not implemented, on #4009

`Access.at/1` can be used to access any index in a list.

```
  ## Examples

      iex> list = [%{name: "john"}, %{name: "mary"}]
      iex> get_in(list, [Access.at(1), :name])
      "mary"
      iex> get_and_update_in(list, [Access.at(0), :name], fn
      ...>   prev -> {prev, String.upcase(prev)}
      ...> end)
      {"john", [%{name: "JOHN"}, %{name: "mary"}]}
      iex> pop_in(list, [Access.at(0), :name])
      {"john", [%{}, %{name: "mary"}]}
```

eksperimental commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,58 @@ defmodule Access do
+      ** (RuntimeError) Access.at/1 expected a list, got: %{}
+  """
+  def at(index) do
+    fn (op, data, next) -> at(op, data, index, next) end
```
please no space after `fn` if you will use `()`

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,58 @@ defmodule Access do
   end
+
+  @doc ~S"""
+  Accesses the _nth_ element (zero based) of a list.
```
It doesn't make sense to say `nth` and zero based. You probably want to say: "Accesses the element at `index` (zero based) of a list".

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,58 @@ defmodule Access do
+
+  defp at([], _index, _next, get, updates) do
+    {get, :lists.reverse(updates)}
+  end
```
Here is a different take that does not require traversing the whole list:

```elixir
  defp at([head | rest], 0, next, updates) do
    case next.(head) do
      {get, update} -> {get, :lists.reverse([update | updates], rest)}
      :pop -> {head, :lists.reverse(updates, rest)}
    end
  end

  defp at([head | rest], index, next, updates) do
    at(rest, index - 1, next, [head | updates])
  end

  defp at([], _index, _next, updates) do
    {get, :lists.reverse(updates)}
  end
```

amuino replied:
💚  no need to keep iterating after we found the index, will update the code.

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,58 @@ defmodule Access do
+      iex> get_in(%{}, [Access.at(1)])
+      ** (RuntimeError) Access.at/1 expected a list, got: %{}
+  """
+  def at(index) do
```
We should probably add a guard that checks the index is `>= 0`.

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,58 @@ defmodule Access do
+  @doc ~S"""
+  Accesses the _nth_ element (zero based) of a list.
+
+  ## Examples
```
We need to document and add examples when the index is out of bounds in both get and update. It seems in update the function is never called... which should be documented.

amuino commented:
Updated the code:
- Improve docs about index being zero-based
- Add docs and examples for invalid indexes. Added a catch-all case for nicer error reporting.
- Alternative implementation for `at/5`, based on @josevalim comments. I added `:get_and_update` as first argument to keep the `at/5` arity, since `at/4` is a different thing.

antipax commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+      {"john", [%{}, %{name: "mary"}]}
+
+  When the index is out of bounds, `nil` is returned and the update function is never called:
+      iex> list = [%{name: "john"}, %{name: "mary"}]
```
Probably add another newline here?

antipax commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+  end
+
+  def at(index) do
+    raise "Access.at/1 expected a non-negative index, got: #{inspect index}"
```
Probably should be `raise ArgumentError, "Access.at/1...".

amuino replied:
I'm keeping this as is since it is the same pattern used for `all`, `elem`…

antipax commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+  end
+
+  defp at(_op, data, _index, _next) do
+    raise "Access.at/1 expected a list, got: #{inspect data}"
```
Same here as above.

antipax replied:
Or for both just remove the catch-alls and allow the no function clause error to do the job. :)

amuino replied:
I was following the same pattern used in `all`, `elem`… in the same file (no explicit ArgumentError, custom error messages).

I could leave it as-is or add the ArgumentError everywhere on the file so that it is consistent… what do you think?

josevalim replied:
This one should be kept as is because it is being invoked inside a callback and not directly a user argument. You can see the rationale above. :)

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+      iex> pop_in(list, [Access.at(0), :name])
+      {"john", [%{}, %{name: "mary"}]}
+
+  When the index is out of bounds, `nil` is returned and the update function is never called:
```
Please add a new line here.

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+  end
+
+  defp at(:get_and_update, data, index, next) when is_list(data) do
+    at(:get_and_update, data, index, next, [])
```
Just give the function another name. For example, `get_and_update_at/3`. :)

amuino replied:
Will do! I was «inspired» by `all` which just uses different arity and thought it was a convention.

josevalim commented on lib/elixir/lib/access.ex:
```diff
@@ -422,4 +422,73 @@ defmodule Access do
+    fn(op, data, next) -> at(op, data, index, next) end
+  end
+
+  def at(index) do
```
There is no need for this clause, in this particular case, it should be obvious in the stacktrace. The reason we add a explicit raise in the other cases is because it is often invoked inside a callback.

amuino commented:
- Removed clause with custom error when index is negative, now raises a standard `FunctionClauseError` (and is documented)
- Whitespace in docs
- Renamed `at/5` → `get_and_update_at/4`
- Not changing `raise "blah blah"` → `raise ArgumentError, "blah blah"` for consistency with existing code (although it seems like a reasonable error type)

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
